Classes/MonomeArc.sc

```diff
@@ -0,0 +1,343 @@
+						addCallback.value(registeredDevices[dvcID], portlst[dvcID], prefixes[dvcID]);
+						seroscnet.sendMsg("/serialosc/notify", "127.0.0.1", NetAddr.localAddr.port);
+					},
+					{
+						("!! no monome arc detected at device slot " ++ devicenum).warn;
+					}
+				);
+			},
+			{
+				("no monome arc detected at device slot " ++ devicenum).warn;
+			}
+		);
+	}
+
+	usePort { arg portnum;
+		dvcID = portlst.indexOf(portnum);
+		oscout = NetAddr.new("localhost", portnum);
+		Post << "MonomeArc: using device # " << dvcID << Char.nl;
+
+		oscout.sendMsg("/sys/port", NetAddr.localAddr.port);
+		oscout.sendMsg("/sys/prefix", prefixID);
+		oscout.sendMsg("/sys/rotation", rot);
+	}
+
+	delta { arg func;
+		deltaFunc = OSCdef.newMatching(
+			("deltaFunc_" ++ dvcID).asSymbol,
+			{ arg message, time, addr, recvPort;
+				var n = message[1], d = message[2];
+				tick[n] = tick[n] + d;
+				if( dvcID.notNil,{
+					if( this.port.value() == addr.port, {
+						if( tick[n].abs >= sens[n], {
+							var val = tick[n]/sens[n];
+							if( val > 0,
+								{val = val.floor},
+								{val = val.ceil}
+							);
+							func.value(n, val);
+							tick[n] = tick[n].mod(sens[n]);
+						});
+					});
+				});
+			},
+			prefixID++"/enc/delta"
+		);
+	}
+
+	key { arg func;
+		keyFunc = OSCdef.newMatching(
+			("keyFunc_" ++ dvcID).asSymbol,
+			{ arg message, time, addr, recvPort;
+				var n = message[1], s = message[2];
+				if( dvcID.notNil,{
+					if( this.port.value() == addr.port, {
+						func.value(n, s);
+					});
+				});
+			},
+			prefixID++"/enc/key"
+		);
+	}
+
+	refresh {
+		for (0, 3, {
+			arg i;
+			if(quadDirty[dvcID][i] != 0,
+				{
+					oscout.sendMsg(
+						prefixID++"/ring/map",
+						i,
+						*ledQuads[dvcID][i]
+					);
+					quadDirty[dvcID][i] = 0;
+				}
+			);
+		});
+
+	}
+
+	led { | ring, led, val |
+		ledQuads[dvcID][ring][(led + scaleFactor).wrap(0, 63)] = val;
+		quadDirty[dvcID][ring] = 1;
+	}
+
+	all { | ring, val |
+		for( 0, 63, {
+			arg i;
+			this.led(ring, i, val);
+		});
+	}
+
+	dark {
+		for( 0, 3, {
+			arg i;
+			this.all(i,0);
+		});
+	}
+
+	ringmap	{ | ring, larr |
+
+		scaleFactor.do({
+
+			larr = larr.shift(1, larr @ (larr.size - 1));
+
+		});
+
+		oscout.sendMsg(prefixID++"/ring/map", ring, *larr);
+
+	}
+
+	segment { | ring, from, to, val |
+		for( from, to, {
+			arg i;
+			this.led(ring, i, val);
+		});
+	}
+
+
+	setSens { | ring, sensitivity |
+		if( sensitivity == 0, {sensitivity = 1});
+		sens[ring] = sensitivity;
+		tick[ring] = 0;
+	}
+
+	// exercise caution when changing rotation
+	// after change, your led positions may not be desirable.
+	setRot { arg degree;
+
+		rot = degree;
+
+		case
+		{ rot == 0 } { scaleFactor = 0 }
+		{ rot == 90 } { scaleFactor = 16 }
+		{ rot == 180 } { scaleFactor = 32 }
+		{ rot == 270 } { scaleFactor = 48 }
+		{ (rot != 0) or: (rot != 90) or: (rot != 180) or: (rot != 270) } {
+
+			"Did not choose valid rotation (0, 90, 180, 270). Using default: 0.".warn;
```

Comment:
i'd suggest simply rounding to the closest valid value and not issuing a warning.